### ISSUE TITLE
goroutines leak in Reauthenticate

### ISSUE THREAD

**Issue by pamburus** [gophercloud/gophercloud#1977]
Following code in provider_client.go does not work as described:
```go
	// Report result to all other interested goroutines.
	//
	// This happens in a separate goroutine because another goroutine might have
	// acquired a copy of `client.reauthmut.ongoing` before we cleared it, but not
	// have come around to sending its request. By answering in a goroutine, we
	// can have that goroutine linger until all responseChannels have been sent.
	// When GC has collected all sendings ends of the channel, our receiving end
	// will be closed and the goroutine will end.
	go func() {
		for responseChannel := range messages {
			responseChannel <- err
		}
	}()
```

Garbage collector does not collect such goroutines and they leak.

Here is a simple example which proves that:
https://play.golang.org/p/99jyHd9R_AX

**Comment by majewsky** [comment]
> Here is a simple example which proves that

This example doesn't prove anything. The goroutine "leak" in this example happens because you keep spawning goroutines on a long-lived channel. This is not the case in the reauth code, there we have short-lived channels and only one goroutine dealing with the channel teardown.

I'm not saying there is no problem with this code. The goroutine leak is probably true, but the presented evidence does not make that clear. I want an actual example code first, then we can verify that #1978 really fixes the issue.

**Comment by majewsky** [comment]
I whipped up a quick real-life testcase to check if there is really a goroutine leak:

```go
package main

import (
	"fmt"
	"runtime"
	"time"

	"github.com/gophercloud/utils/openstack/clientconfig"
)

func main() {
	pc, err := clientconfig.AuthenticatedClient(nil)
	must(err)
	_ = pc

	for iteration := 0; iteration < 20; iteration++ {
		must(pc.Reauthenticate(pc.Token()))
		fmt.Printf("after % 3d reauths: %d goroutines\n", iteration, runtime.NumGoroutine())

		//don't put too much load on Keystone
		time.Sleep(1 * time.Second)
	}
}

func must(err error) {
	if err != nil {
		panic(err.Error())
	}
}
```

Turns out there isn't:

```
$ go run main.go
after   0 reauths: 2 goroutines
after   1 reauths: 2 goroutines
after   2 reauths: 2 goroutines
after   3 reauths: 2 goroutines
after   4 reauths: 2 goroutines
after   5 reauths: 2 goroutines
after   6 reauths: 2 goroutines
after   7 reauths: 2 goroutines
after   8 reauths: 2 goroutines
after   9 reauths: 2 goroutines
after  10 reauths: 2 goroutines
after  11 reauths: 2 goroutines
after  12 reauths: 2 goroutines
after  13 reauths: 2 goroutines
after  14 reauths: 2 goroutines
after  15 reauths: 2 goroutines
after  16 reauths: 2 goroutines
after  17 reauths: 2 goroutines
after  18 reauths: 2 goroutines
after  19 reauths: 2 goroutines
after  20 reauths: 2 goroutines
```

**Comment by pamburus** [comment]
This code does nothing, so it has no leaks.
It has no AllowReauth option.
So, `pc` has not `ReauthFunc`.
So, `go func` is never executed.

Do not use `clientconfig.AuthenticatedClient`.
Use `openstack.NewClient` and `openstack.Authenticate` with `AllowReauth` set to `true`.
Also, add some `fmt.Println("goroutine started")` into the function which is executed in goroutine.

I tried that with your example code, and it leaks.

**Comment by majewsky** [comment]
Thanks for catching that. This testcase reproduces the issue:

```go
package main

import (
        "fmt"
        "runtime"
        "sync"
        "time"

        "github.com/gophercloud/gophercloud/openstack"
        "github.com/gophercloud/utils/openstack/clientconfig"
)

func main() {
        opts, err := clientconfig.AuthOptions(nil)
        must(err)
        opts.AllowReauth = true
        pc, err := openstack.AuthenticatedClient(*opts)
        must(err)

        for iteration := 0; iteration < 20; iteration++ {
                var wg sync.WaitGroup
                wg.Add(10)
                for branch := 0; branch < 10; branch++ {
                        go func() {
                                defer wg.Done()
                                must(pc.Reauthenticate(pc.Token()))
                        }()
                }
                wg.Wait()
                fmt.Printf("after % 3d reauths: %d goroutines\n", iteration, runtime.NumGoroutine())

                //don't put too much load on Keystone
                time.Sleep(1 * time.Second)
        }
}

func must(err error) {
        if err != nil {
                panic(err.Error())
        }
}
```

**Comment by jtopjian** [comment]
Fixed in #1978
